### DISCUSSION TITLE
NaN in monitor console row with RWALL block

### DISCUSSION THREAD

**Discussion by rzadki22** [OpenRadioss/OpenRadioss#2082] [Q&A]
Hello, 

I noticed weird behavior of engine during calculation of a free fall with elements deletion when object hits to a rigid wall. After some increments a row with NaN occurs, after that solver does not return errors but go further and saves empty animation files and finishes work without any exceptions : P
Here you can find snippet from the out file:

```
   50700  0.7065E-02  0.1393E-06 NODE       21133 -25.8%   175.1       3532.       0.000       38.73      0.1086E-01  0.4215E-03  0.4526E-05
   50800  0.7079E-02  0.1393E-06 NODE       27147   0.0%         NaN         NaN   0.000             NaN  0.1086E-01  0.4215E-03  0.4526E-05
     ANIMATION FILE: B_Plate_Free_FallA072 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA073 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA074 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA075 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA076 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA077 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA078 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA079 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA080 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA081 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA082 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA083 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA084 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA085 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA086 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA087 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA088 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA089 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA090 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA091 WRITTEN
     ANIMATION FILE: B_Plate_Free_FallA092 WRITTEN

     RESTART FILES: B_Plate_Free_Fall_0001_[0001-0001].rst WRITTEN
```
After some trials I noticed that this happens when (or not) I change parameters in a RWALL block. For example with friction coefficient 0.05 it does not work, but with 0.01 it does, but with 0 again not and so on : )

I think that something should be improved in code to catch the exception of this misbehavior.

Best regards, 
Filip

This is my first post here and I see a lot of opportunity for usage OpenRadioss for better engineering : ) If you need any details do not hesitate to contact.

**Comment by Schwoertzig** [comment]
Hi @rzadki22 
thank you for your interest in OpenRadioss and for reporting this possibility to improve. You are right whenever NaN appears something can be improved. 

- If you would like to investigate and submit an enhancement I can help you. You will need to compile and run a debug version of OpenRadioss.
- If you don't feel comfortable with it I can also do it for you. Is it possible to share the input file which leads to the NaN ?

Regards

**Reply by roguic** [reply]
Hi @rzadki22 ,

To trap the Nan, you can add the following line to your *1.rad :
/DEBUG/NAN

The computation will stop if a Nan is found in the acceleration array ( A array in the code) and some information will be printed.

Best,
Romain

**Reply by rzadki22** [reply]
Hi @Schwoertzig - thank you for answer. ad.1 - Do you mean special version for debugging or usual one but compiled? I have compiled version (OpenRadioss-latest-20231027) on my Win right now. ad.2 - sure, I can share files - it is not confidential. What is the best way to share files, it weighs around 8MB.

Hi @roguic - thank you for your answer - I will try it soon and share the output in this discussion : )

Best regards,
Filip

**Reply by rzadki22** [reply]
Hi @roguic ,
with /DEBUG/NAN I got not so more information in "*1.out":


```
...  
 45900  0.6396E-02  0.1393E-06 NODE       15637 -27.9%   201.4       3266.       0.000      -147.5      0.1107E-01  0.4216E-03  0.4614E-05
     ANIMATION FILE: B_Plate_Free_FallA065 WRITTEN
   46000  0.6410E-02  0.1393E-06 NODE        1937 -27.9%   201.7       3265.       0.000      -147.5      0.1107E-01  0.4216E-03  0.4614E-05
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      24953 Global node ID:     27534
   A values:        NaN       NaN       NaN
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      24954 Global node ID:     27537
   A values:        NaN       NaN       NaN
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      24955 Global node ID:     27535
   A values:        NaN       NaN       NaN
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      24956 Global node ID:     27538
   A values:        NaN       NaN       NaN
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      25013 Global node ID:     27579
   A values:        NaN       NaN       NaN
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      25014 Global node ID:     27582
   A values:        NaN       NaN       NaN
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      25015 Global node ID:     27580
   A values:        NaN       NaN       NaN
   NaN detected on A array on processor ID         0 at cycle     46003
   Local node ID:      25016 Global node ID:     27583
   A values:        NaN       NaN       NaN

   ********************* NaN ********************* 
   ******** I"m going to kill the run now ******** 
   ********************* NaN ********************* 

     ANIMATION FILE: B_Plate_Free_FallA066 WRITTEN
...
```

**Comment by Schwoertzig** [comment]
Hi @rzadki22 ,
I would recommend to add this when compiling debug version : 
-  -addflag="-ffpe-trap=invalid" 
It is supposed to send a specific signal when NaN occurs. This signal can be caught by the debugger and you will spot the exact location in the source code.
_Please note that if you already compiled a debug version, you may need to clean the previous compilation  ( -clean )  and restart with this new flag._

**Reply by rzadki22** [reply]
Hi @Schwoertzig ,
thank you for hints - i compiled new starter and engine *_db.exe  with ffpe flag as you wrote above but unfortunately something I am doing wrong and my output is not more verbose than I had written earlier. If you like to check this issue on your own, please find input files in the attachment.

[input_files_nan.zip](https://github.com/OpenRadioss/OpenRadioss/files/13383147/input_files_nan.zip)

Best regards and I am looking forward for your comments : )
Filip

**Reply by Schwoertzig** [reply]
Hi @rzadki22 ,
I found an invalid operation in  mmain > m1lawtot > m1ismstr10 > princ_u.
 I am going to check how to fix it.
Thierry

**Reply by Schwoertzig** [reply]
Hi @rzadki22 ,
There is an input issue in your input file B_Plate_Free_Fall_0000.rad .
Failure criterion /FAIL/TENSSTRAIN has incorrect fileds. eps_t1 must be in column 1-2 and eps_t2 in column 3-4
![image](https://github.com/OpenRadioss/OpenRadioss/assets/104987175/47b49d9a-9fa8-46f6-9860-e4a36f7aea3d)

In addition to that you can notice than eps_t1 < eps_t2. this should be detected as an error by Starter once it is correctly read. In order to check that it is correctly read, you can have a look in Starter listing file : B_Plate_Free_Fall_0000.out 
![image](https://github.com/OpenRadioss/OpenRadioss/assets/104987175/af08a3d6-67ba-46e9-9255-81c9f3d7b3a1)

The NaN is coming from material law 1 when element have a very bad shape. (negative Jacobian).
![image](https://github.com/OpenRadioss/OpenRadioss/assets/104987175/1f7daffc-9b6b-43be-ba80-0c20993c2d80)
You may first run the simulation again with the expected failure criteria (there are currently a lot of element deletion) and check if the bad elem quality still occur. 

Regards

**Reply by Schwoertzig** [reply]
@rzadki22 ,
The encountered NaN is related to total Strain formulation once the element is deleted. @qiangRadioss will improve the behavior for related elem formulations.
You can launch your simulation with **Ismstr=4**  (Since you have /DEF_SOLID option you can simply  update its flag ISMSTR from -2 to -1)
Regards

**Reply by rzadki22** [reply]
@Schwoertzig

thank you for you explanation - I checked manual reference about TENSSTRAIN and it's clear now.
One thing about way of how the starter works is unclear for me though. 
In my case with colums corresponding to eps_t have two float values and columuns assigned to eps_t2 include no input (default 2e30?). Why did the starter allow two floats in the block with one permissible input figure?
It is unclear for me why eps_t2 equals 0.5e-3 if the default value acc.to the manual equals 2e30.

Regards
Filip

**Reply by Schwoertzig** [reply]
Hi @rzadki22 ,
The rules are the following ones : 

- 10 columns per line
- 10 digits per column (100 per line)
- integer are always input within a single column
- real/float are always input with 2 columns

eps_t1="    0.0001   0.00005" was read as eps_t2="0.0001"and "                     " was read as 0.0 and set to default value 2e20  (eps_t1 < eps_t2). I don't know why the Starter was able to read something when you had 2 float values within 2 columns only (which is unexpected). This is a good question. I am going to check if this behavior may be improved.

Regards

**Reply by rzadki22** [reply]
Hi @Schwoertzig, 
You write that an eps_t2 shall be default in our example but the .out snippet shows 5e-4. Does starter assume and replace some value higher than eps_t1? like 5 * eps_t1?
Regards

**Reply by Schwoertzig** [reply]
With the initial input which was wrong :

```
/FAIL/TENSSTRAIN/1
#---eps_t|--eps_t2-|--fct_ID-----------|--fct_ID-|----eps_f1---------|--------eps_f2-----|--S-flag-|
    0.0001   0.00005                                            0.0008              0.0004         1
```

it leads to eps_t2 = 0. (default = 2.e20) since Starter does not read any value in columns 3-4

```
         TENSION STRAIN FAILURE MODEL        
         ------------------------            
     MAXIMUM TENSION STRAIN 1. . . . . . . . . . . . . . . . =  0.1000E-03
     MAXIMUM TENSION STRAIN 2. . . . . . . . . . . . . . . . =  0.2000E+21
     MAXIMUM STRAINS SCALING FUNCTION. . . . . . . . . . . . =         0
     FAILURE STRAIN INPUT TYPE . . . . . . . . . . . . . . . =         1
     ...
     MAXIMUM FIRST PRINCIPAL STRAIN FOR ELEMENT DELETION . . =  0.8000E-03
     MAXIMUM SECOND PRINCIPAL STRAIN FOR ELEMENT DELETION. . =  0.4000E-03
```

If you fix the alignment for eps_t2 :
```
/FAIL/TENSSTRAIN/1
#-------------eps_t1|------------eps_t2|--fct_ID-|----eps_f1---------|--------eps_f2-----|--S-flag-|
               0.0001            0.00005                        0.0008              0.0004         1
```

then Starter reads correctly your value eps_t2=5e-4 (columns 3-4) and print an error message since 0.0001 > 0.00005

```
ERROR ID :    617
** ERROR IN FAILURE MODEL DEFINITION
DESCRIPTION :  
   -- MATERIAL ID: 1
   EPST1 MUST BE LESS THAN EPST2

         TENSION STRAIN FAILURE MODEL        
         ------------------------            
     MAXIMUM TENSION STRAIN 1. . . . . . . . . . . . . . . . =  0.1000E-03
     MAXIMUM TENSION STRAIN 2. . . . . . . . . . . . . . . . =  0.5000E-04

```

Engine will not be able to be launched with such an error. So you must update the value eps_t2  with the expected one (such as eps_t1 < eps_t2)

Does it answer to your question ?
Regards

**Reply by rzadki22** [reply]
Hi @Schwoertzig,

thank you for a wider expIanation, I was little confused of previous snippets.

Regards
